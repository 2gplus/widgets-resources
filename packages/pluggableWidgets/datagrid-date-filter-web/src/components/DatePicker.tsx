import { createElement, Dispatch, Fragment, MutableRefObject, ReactElement, SetStateAction, useState } from "react";
import DatePickerComponent from "react-datepicker";
import classNames from "classnames";
import { isDate, isValid } from "date-fns";

interface DatePickerProps {
    adjustable: boolean;
    dateFormat?: string;
    locale?: string;
    name?: string;
    placeholder?: string;
    screenReaderInputCaption?: string;
    setValue: Dispatch<SetStateAction<Date | null>>;
    value: Date | null;
    ref?: MutableRefObject<DatePickerComponent | null>;
}

export function DatePicker(props: DatePickerProps): ReactElement {
    const [open, setOpen] = useState(false);
    return (
        <Fragment>
            <span className="sr-only" id={`${props.name}-label`}>
                {props.screenReaderInputCaption}
            </span>
            <DatePickerComponent
                allowSameDay={false}
                ariaLabelledBy={`${props.name}-label`}
                autoFocus={false}
                className={classNames("form-control", { "filter-input": props.adjustable })}
                dateFormat={props.dateFormat}
                disabledKeyboardNavigation={false}
                dropdownMode="select"
                locale={props.locale}
                onChange={date => {
                    if (isDate(date) && isValid(date)) {
                        props.setValue(date as Date);
                    } else {
                        props.setValue(null);
                    }
                }}
                open={open}
                onClickOutside={() => setOpen(false)}
                placeholderText={props.placeholder}
                popperPlacement="bottom-start"
                popperModifiers={{
                    offset: {
                        enabled: true,
                        offset: "0, 0"
                    },
                    preventOverflow: {
                        enabled: true,
                        escapeWithReference: false,
                        boundariesElement: "viewport"
                    },
                    flip: {
                        enabled: false
                    }
                }}
                preventOpenOnFocus
                ref={props.ref}
                selected={props.value}
                shouldCloseOnSelect={false}
                showMonthDropdown
                showPopperArrow={false}
                showYearDropdown
                strictParsing
                useWeekdaysShort
            />
            <button className="btn btn-default btn-calendar" onClick={() => setOpen(true)}>
                <span className="glyphicon glyphicon-calendar" />
            </button>
        </Fragment>
    );
}
