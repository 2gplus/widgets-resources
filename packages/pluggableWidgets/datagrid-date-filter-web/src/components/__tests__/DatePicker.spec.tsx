import { render, shallow } from "enzyme";
import { createElement } from "react";
import { DatePicker } from "../DatePicker";
import ReactDOM from "react-dom";

describe("Date picker component", () => {
    beforeAll(() => {
        jest.spyOn(global.Math, "random").mockReturnValue(0.123456789);

        // @ts-ignore
        jest.spyOn(ReactDOM, "createPortal").mockReturnValue((element, node) => {
            return element;
        });
    });

    it("renders correctly", () => {
        const component = render(
            <DatePicker adjustable value={null} setValue={jest.fn()} dateFormat="dd/MM/yyyy" locale="nl-NL" />
        );

        expect(component).toMatchSnapshot();
    });

    it("renders correctly when is not adjustable", () => {
        const component = render(
            <DatePicker adjustable={false} value={null} setValue={jest.fn()} dateFormat="dd/MM/yyyy" locale="nl-NL" />
        );

        expect(component).toMatchSnapshot();
    });

    it("renders correctly with different locale and date format", () => {
        const component = render(
            <DatePicker adjustable={false} value={null} setValue={jest.fn()} dateFormat="yyyy-MM-dd" locale="pt-BR" />
        );

        expect(component).toMatchSnapshot();
    });

    it("calls for setValue when value changes", () => {
        const setValue = jest.fn();
        const component = shallow(
            <DatePicker adjustable value={null} setValue={setValue} dateFormat="dd/MM/yyyy" locale="nl-NL" />
        );

        component.find("r").simulate("change", { target: { value: "01/12/2020" } });

        expect(setValue).toBeCalledTimes(1);
    });
});
