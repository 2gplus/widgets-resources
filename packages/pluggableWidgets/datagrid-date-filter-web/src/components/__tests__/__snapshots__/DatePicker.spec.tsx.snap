// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date picker component renders correctly 1`] = `
Array [
  <span
    class="sr-only"
    id="undefined-label"
  />,
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container"
    >
      <input
        aria-labelledby="undefined-label"
        class="form-control filter-input"
        type="text"
        value=""
      />
    </div>
  </div>,
  <button
    class="btn btn-default btn-calendar"
  >
    <span
      class="glyphicon glyphicon-calendar"
    />
  </button>,
]
`;

exports[`Date picker component renders correctly when is not adjustable 1`] = `
Array [
  <span
    class="sr-only"
    id="undefined-label"
  />,
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container"
    >
      <input
        aria-labelledby="undefined-label"
        class="form-control"
        type="text"
        value=""
      />
    </div>
  </div>,
  <button
    class="btn btn-default btn-calendar"
  >
    <span
      class="glyphicon glyphicon-calendar"
    />
  </button>,
]
`;

exports[`Date picker component renders correctly with different locale and date format 1`] = `
Array [
  <span
    class="sr-only"
    id="undefined-label"
  />,
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container"
    >
      <input
        aria-labelledby="undefined-label"
        class="form-control"
        type="text"
        value=""
      />
    </div>
  </div>,
  <button
    class="btn btn-default btn-calendar"
  >
    <span
      class="glyphicon glyphicon-calendar"
    />
  </button>,
]
`;
